Allow only one trusted executor producing the execution receipt

liuchengxu opened the pull request:
With this PR, only one single executor authority is allowed to generate and publish the execution receipt for each primary block, which must be initialized at genesis, hard coded as Alice for now.

Apart from the known issue of running multi executors, I found that the codec of OpaqueExecutionReceipt and ExecutionReceipt is broken since some commit, now the encoding of OpaqueExecutionReceipt and ExecutionReceipt are not compatible with raw Vec<u8> encoding, I'll fix it in a later PR. There are also a few TODOs like we should support the executor setting a new authority key, but we can do it later as well. Reviewing commit by commit is recommended.

liuchengxu committed:
Replace the trait alias for primary chain interface with generic

liuchengxu committed:
Introduce Authority in pallet-executor

There is only one trusted executor in the initial version.

liuchengxu committed:
Expose executor authority id via runtime api

liuchengxu committed:
Generate execution receipt when current executor authority key exists in the keystore

liuchengxu committed:
Introduce SignedExecutionReceipt

liuchengxu committed:
Gossip SignedExecutionReceipt over the executor network

So that the other executors can verify the execution receipt author.

liuchengxu committed:
Verify the authority of external execution receipt

liuchengxu committed:
Add the executor authority key into keystore in dev mode

Specify `--alice --dev` in the secondary chain args, the executor
authority key(Alice) will be added to the keystore automatically.
```bash
./target/release/subspace-node --dev -d tmp --log=txpool=trace,gossip::executor=trace \
  -- --alice --dev --port 40333 --log=cirrus::executor=trace,txpool=trace,gossip=trace --rpc-port 8845 --ws-port 8846
```

nazar-pc commented on crates/sp-executor/src/lib.rs:
```diff
@@ -30,6 +31,22 @@ use sp_trie::StorageProof;
 use subspace_runtime_primitives::{AccountId, Hash as PHash};
 
+/// Key type for Executor.
+pub const KEY_TYPE: KeyTypeId = KeyTypeId(*b"exec");
```
Let's not make this public, it doesn't need to be.

vedhavyas commented on cumulus/client/cirrus-executor/src/lib.rs:
```diff
@@ -663,15 +668,34 @@ where
+			.runtime_api()
+			.authority_id(&BlockId::Hash(primary_hash))?;
+		if *signer != expected_authority_id {
+			// TODO: handle the misbehavior.
```
can we block the executor on the libp2p layer itself for misbehaving?

nazar-pc replied:
I _think_ this error should contribute to node's reputation already, but someone needs to double-check it with source code.

liuchengxu replied:
Actually, I recall that Jeremiah is not a fan of the peer reputation system in Substrate. Anyway, we'll discuss this before proceeding.

nazar-pc commented on crates/pallet-executor/src/lib.rs:
```diff
@@ -169,6 +170,35 @@ mod pallet {
+    /// A tuple of (stable_executor_id, executor_signing_key).
+    #[pallet::storage]
+    #[pallet::getter(fn authority)]
+    pub(super) type Authority<T: Config> =
```
Wouldn't `Executor` be a better name than `Authority` here and in other places?

liuchengxu replied:
Hmm, not quite sure about that. `Authority` also means that the executors are not permissionless like farmers.

liuchengxu replied:
No strong opinion though, if you do think `Executor` is a better one, I'm totally fine with it.

nazar-pc replied:
I think it makes sense to use `Executor` and `ExecutorId` even though we have other `Executor` structs that I'm not a big fan of, but what can we do about it :slightly_smiling_face:

liuchengxu replied:
I don't have a better answer, just updated as requested, hope that's what you have in your mind. We'll iterate this quickly anyway.

nazar-pc commented on crates/subspace-runtime/src/lib.rs:
```diff
@@ -1048,6 +1048,12 @@ impl_runtime_apis! {
+        fn authority_id() -> sp_executor::AuthorityId {
+            Executor::authority()
+                .map(|(_account_id, authority_id)| authority_id)
+                .expect("Executor authority must be provided; qed")
```
Maybe it shouldn't be `Option` then and we can avoid this?

liuchengxu replied:
This storage item has to be an `OptionQuery` because `T::AccountId` has removed the Default bound.

nazar-pc replied:
I see. I guess this will go away as we change this to vector or some set/map in the future.

liuchengxu replied:
Yes, indeed.

nazar-pc commented on cumulus/client/cirrus-executor/src/processor.rs:
```diff
@@ -206,17 +210,43 @@ where
 
 		// The applied txs can be fully removed from the transaction pool
```
BTW, is this comment a TODO? It doesn't seem to point to something specific.

liuchengxu replied:
Yes, a TODO, will update it!

nazar-pc commented on cumulus/client/cirrus-executor/src/processor.rs:
```diff
@@ -206,17 +210,43 @@ where
+					// Return `Some(_)` to broadcast ER to all farmers via unsigned extrinsic.
+					Ok(Some(execution_receipt.into()))
+				},
+				Ok(None) => Err(sp_blockchain::Error::Application(Box::from(
```
`Box::new`?

liuchengxu replied:
`Box::new` can't be used as it requires StdError which is not implemented by String and str according to the compiler.

nazar-pc commented on cumulus/client/cirrus-executor/src/processor.rs:
```diff
@@ -246,9 +246,9 @@ where
+						tracing::error!(target: LOG_TARGET, error = ?e, "Failed to send signed execution receipt");
 					}
 
 					// Return `Some(_)` to broadcast ER to all farmers via unsigned extrinsic.
```
Why below ER is not signed even though comment says it needs to be broadcast to other executors, isn't it what the channel above does?

liuchengxu replied:
The comment said it will be broadcasted to all **farmers** via unsigned extrinsic, farmers will see `ExecutionReceipt` from the unsigned extrinsic, executors will see `SignedExecutionReceipt` sent over the channel from the executor gossip network.

nazar-pc replied:
But farmers also need to have signed extrinsic, don't they?

liuchengxu replied:
I don't understand why farmers need to have signed extrinsic. We have no signed extrinsic on this topic so far.

nazar-pc replied:
How would farmers know that ER is produced by allowed executor and not by some random person on the Internet?

liuchengxu replied:
That's one thing I'd love to discuss along with the above TODO of execution receipt signing misbehavior handling. Farmers can not verify the validity of the execution receipt signature of a previous block within the runtime because runtime can not access the info of some old block. Another TODO :P https://github.com/subspace/subspace/blob/e848e9e153e84fa90c6e5eddd1cfabee63cdece6/crates/pallet-executor/src/lib.rs#L85

nazar-pc replied:
Farmers should be able to verify validity because they have access to executor set and signature must be produced by one of executors that are designated to produce ER at this block height.

nazar-pc replied:
Especially now when we have just one executor

liuchengxu replied:
Well, it should verify the validity for sure, the problem is how we implement it, as I mentioned, this verification can't be done in the runtime directly, which is similar to the fraud proof verification problem.

nazar-pc replied:
We definitely can't verify the contents, but we don't need that anyway unless we have a fraud proof. But we can and need to verify the signature so at least we know it is produced by properly staked executor.

liuchengxu replied:
As I have said, you can't do it in runtime directly but have to verify it on the client side as you can't access the executor set of a previous block within the runtime(the execution receipt belongs to a previous block). Then the problem becomes if we're going to use the externality approach or a complete off-runtime approach. Personally, I second the externality way as this verification is relatively effortless.

nazar-pc replied:
Now I see the problem better, thanks for clarifying. I don't think this is unsolvable though. If we update validator set at the end of the block (after checking all of the extrinsics, including ERs), we don't need to reach out to the previous block at all. Also we can keep 2 sets in the state potentially.

I think this is very important and needs to change (even if we don't validate the transaction in the runtime, it still needs to be signed so the client can check it), but we can postpone it to the next PR.

liuchengxu replied:
The previous block does not mean the just last block, possibly some old block which is a few blocks earlier as you can't guarantee the receipt of the current block will be included in the next block? Additionally, if we have another way which is not complicated in my opinion, why do we have to unnecessarily add more state in the runtime?

Yeah, we'll discuss a few things before moving forward.

nazar-pc commented on cumulus/runtime/src/runtime.rs:
```diff
@@ -124,7 +124,10 @@ pub mod opaque {
-	pub struct SessionKeys { }
+	pub struct SessionKeys {
+		/// Primarily used for adding the executor authority key into the keystore in the dev mode.
+		pub executor: sp_executor::ExecutorKey,
```
Where is this actually used?

liuchengxu replied:
https://github.com/paritytech/substrate/blob/a5cd3852d144289777dd4b914b6b3aa7d13bcd85/client/service/src/builder.rs#L483-L488 In Substrate, the session keys will be initialized using the dev seed while starting the node, these initialized keys will be stored in the keystone, so that you don't have to insert them into the keystore manually, facilitating the local development.

nazar-pc replied:
I mean I haven't seen `SessionKeys.executor` field accessed anywhere in the code, but you have just added it. That is why I was confused where is this field actually used. Or does the name not matter at all?

liuchengxu replied:
No, the name of this field does not matter, what matters is the type of this field that implements `BoundToRuntimeAppPublic` which will tell the authority key type.

nazar-pc replied:
OK, makes sense, thanks for explanation!

We should probably remove the `subspace` field from `SessionKeys` in primary chain then as we don't use keychain there for anything.

nazar-pc reviewed (commented):
Some comments and questions

liuchengxu committed:
Apply code review

liuchengxu committed:
Generate execution receipt only when running exeuctor with --validator and the signing key exists in keystore

I think it still makes sense to add this condition as long as we still have the
full node concept for executors.

liuchengxu committed:
Use Executor/ExecutorId instead of Authority/AuthorityId for executors

Fixes https://github.com/subspace/subspace/pull/384#discussion_r855377155

liuchengxu committed:
Merge branch 'main' into trusted-executor
